### PR TITLE
[Fix] `no-unknown-property`: properly recognise unknown HTML/DOM attributes

### LINKED ISSUE #757: `no-unknown-property` doesn't prevent use of unknown properties?
Today React gave me an ["unknown prop"](https://facebook.github.io/react/warnings/unknown-prop.html) warning:

image
In this example, I gave my `div` element an `abc` attribute, which is obviously invalid HTML.

I was surprised that my linter didn't pick up on this. Looking at the documentation, the name and description of the `no-unknown-property` rule lead me to believe that this should be caught, but that doesn't appear to be the case.

So, I suppose this is a feature request: add a rule to catch invalid DOM properties.

### PR DESCRIPTION
**Pull request by sjarva**
Closes #757 by...

- renaming an old error message (`unknownProp` to `unknownPropWithStandardName`), since it reports an unknown property that is close (one lower/higher case letter) to an allowed (standard) name
- adding a new error message (`unknownProp`) that is used when an unknown property (without a 
- listing all one word and multiple word valid HTML/DOM attributes, and `aria-*` attributes
- adding a function that checks if attribute name is a valid `data-*` attribute
- adding more unit test cases
- moving one test that should have failed, from valid to invalid tests
- updating rule documentation

What was not implemented in this PR:
- mapping element specific tags to `ATTRIBUTE_TAGS_MAP` (e.g. attributes `optimum`, `high` and `low` should be used only on `<meter>` element)

While implementing these changes, I felt like I cannot be sure that this PR catches all valid HTML/DOM attributes. There are some listing [on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes) and [on React documentation](https://reactjs.org/docs/dom-elements.html#all-supported-html-attributes), but I could still find attributes that weren't on these listing that I felt would be beneficial to be listed in this rule (for example, the Safari/Apple related `autoCorrect` and `autoSave`, and React related `ref` and `defaultValue`). So, I'm sure there will be cases that this PR has missed, and the community will find them if these changes are merged and published.

Feel free to tag some who knows HTML/DOM attributes well and can help with reviewing and possibly suggesting more new unit test cases.

### CODE REVIEW

File: tests/lib/rules/no-unknown-property.js
Comment (ljharb) [inline tests/lib/rules/no-unknown-property.js:37]: can you elaborate on these?

**Review by ljharb — APPROVED** [review]
LGTM pending one comment

File: tests/lib/rules/no-unknown-property.js
Comment (sjarva) [reply tests/lib/rules/no-unknown-property.js:37]: Sure! I have moved them...

File: tests/lib/rules/no-unknown-property.js
Comment (sjarva) [inline tests/lib/rules/no-unknown-property.js:53]: ...here.

**Comment by ljharb** [comment]
Thanks!

**Comment by sjarva** [comment]
I'm commenting here, because I think this is the best place to do it. I'm aware of the issues and errors that these changes caused. When making this PR, I was sure that I have missed some common scenarios that the community would find, and I was wondering should I ask or comment that could you @ljharb as a maintainer run some additional checks (like the smoke test) before releasing this. Being still a new contributor, I was shy and didn't dare to ask. Looking back now, I should have asked, and I will so better in the future.

Also, I am on another timezone than you are, so I wasn't awake when you released this and tagged me to the first issue. I'm happy that someone else from the community provided a quick fix 🛠️ 🤗

**Comment by valscion** [comment]
Hi! I want to say huge thanks for writing this change — it spotted some cases in our codebase that we would've never gotten to find without the help of this linter!

I figured it's nice to also hear some encouraging words and not merely bug reports :smile: I for one am thankful for the work you've done here 💞

**Comment by sjarva** [comment]
> Hi! I want to say huge thanks for writing this change — it spotted some cases in our codebase that we would've never gotten to find without the help of this linter!
> 
> I figured it's nice to also hear some encouraging words and not merely bug reports 😄 I for one am thankful for the work you've done here 💞

Thank you so much @valscion! Yes, it is nice to hear that it helped find bugs 🐛 💕 And I really appreciate you taking the time to write this comment, it really made my day 🤗

**Pull request merged**
